File: android/app/src/main/java/com/knowyourphone/app/network/Protocol.kt
package com.knowyourphone.app.network

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.annotations.SerializedName
import com.knowyourphone.app.model.HighlightTarget
import com.knowyourphone.app.privacy.RedactionSummary

// --- Client -> Server messages ---

data class AudioDataMessage(
    val type: String = "audio_data",
    val format: String = "wav",
    val sampleRate: Int = 16000
)

data class DeviceInfo(
    val manufacturer: String,
    val model: String,
    val androidVersion: String
)

data class HelloMessage(
    val type: String = "hello",
    val sessionId: String,
    val clientId: String? = null,
    val deviceInfo: DeviceInfo
)

data class SetLanguageMessage(
    val type: String = "set_language",
    val languageCode: String
)

data class ScreenshotResponseMessage(
    val type: String = "screenshot_response",
    val screenshot: String, // base64 JPEG
    val uiTree: JsonObject,
    val redacted: Boolean = false,
    val redactions: List<RedactionSummary> = emptyList()
)

data class ScreenshotDeclinedMessage(
    val type: String = "screenshot_declined"
)

data class CancelMessage(
    val type: String = "cancel"
)

data class SetAutoScreenshotMessage(
    val type: String = "set_auto_screenshot",
    val enabled: Boolean
)

// --- Server -> Client messages ---

sealed class ServerMessage {
    data class Transcript(val text: String) : ServerMessage()
    data class NeedScreenshot(val reason: String) : ServerMessage()
    data class ScreenshotRequest(val text: String, val reason: String, val hasAudio: Boolean) : ServerMessage()
    object AnswerStart : ServerMessage()
    data class Highlights(val highlights: List<HighlightTarget>) : ServerMessage()
    data class AnswerEnd(val text: String, val highlights: List<HighlightTarget>) : ServerMessage()
    data class Error(val message: String) : ServerMessage()
    object Cancelled : ServerMessage()
}

object MessageParser {
    private val gson = Gson()

    fun parseServerMessage(json: String): ServerMessage? {
        return try {
            val obj = gson.fromJson(json, JsonObject::class.java)
            when (obj.get("type")?.asString) {
                "transcript" -> ServerMessage.Transcript(
                    text = obj.get("text")?.asString ?: ""
                )
                "need_screenshot" -> ServerMessage.NeedScreenshot(
                    reason = obj.get("reason")?.asString ?: ""
                )
                "screenshot_request" -> ServerMessage.ScreenshotRequest(
                    text = obj.get("text")?.asString ?: "",
                    reason = obj.get("reason")?.asString ?: "",
                    hasAudio = obj.get("hasAudio")?.asBoolean ?: false
                )
                "answer_start" -> ServerMessage.AnswerStart
                "highlights" -> {
                    val highlights = mutableListOf<HighlightTarget>()
                    obj.getAsJsonArray("highlights")?.forEach { elem ->
                        val h = gson.fromJson(elem, HighlightTarget::class.java)
                        highlights.add(h)
                    }
                    ServerMessage.Highlights(highlights = highlights)
                }
                "answer_end" -> {
                    val highlights = mutableListOf<HighlightTarget>()
                    obj.getAsJsonArray("highlights")?.forEach { elem ->
                        val h = gson.fromJson(elem, HighlightTarget::class.java)
                        highlights.add(h)
                    }
                    ServerMessage.AnswerEnd(
                        text = obj.get("text")?.asString ?: "",
                        highlights = highlights
                    )
                }
                "error" -> ServerMessage.Error(
                    message = obj.get("message")?.asString ?: "Unknown error"
                )
                "cancelled" -> ServerMessage.Cancelled
                else -> null
            }
        } catch (e: Exception) {
            null
        }
    }

    fun <T> toJson(obj: T): String = gson.toJson(obj)
}

File: backend/src/protocol.ts
// WebSocket protocol message types

export interface Highlight {
  elementId: string;
  label: string;
  bounds: {
    left: number;
    top: number;
    right: number;
    bottom: number;
  };
}

// Client -> Server messages

export interface AudioDataMessage {
  type: "audio_data";
  format: "wav";
  sampleRate: number;
}

export interface HelloMessage {
  type: "hello";
  clientId: string;
}

export interface ResetSessionMessage {
  type: "reset_session";
}

export interface SetLanguageMessage {
  type: "set_language";
  languageCode: string;
}

export interface ScreenshotResponseMessage {
  type: "screenshot_response";
  screenshot: string; // base64
  uiTree: UiTree;
}

export interface ScreenshotDeclinedMessage {
  type: "screenshot_declined";
}

export interface CancelMessage {
  type: "cancel";
}

export type ClientMessage =
  | AudioDataMessage
  | HelloMessage
  | ResetSessionMessage
  | SetLanguageMessage
  | ScreenshotResponseMessage
  | ScreenshotDeclinedMessage
  | CancelMessage;

// Server -> Client messages

export interface TranscriptMessage {
  type: "transcript";
  text: string;
}

export interface NeedScreenshotMessage {
  type: "need_screenshot";
  reason: string;
}

export interface ScreenshotRequestMessage {
  type: "screenshot_request";
  text: string;
  reason: string;
  hasAudio: boolean;
}

export interface AnswerMessage {
  type: "answer";
  text: string;
  highlights: Highlight[];
  hasAudio: boolean;
}

export interface SessionStatusMessage {
  type: "session_status";
  sessionId: string;
  userCount: number;
  assistantCount: number;
}

export interface ErrorMessage {
  type: "error";
  message: string;
}

export interface CancelledMessage {
  type: "cancelled";
}

export type ServerMessage =
  | TranscriptMessage
  | NeedScreenshotMessage
  | ScreenshotRequestMessage
  | AnswerMessage
  | SessionStatusMessage
  | ErrorMessage
  | CancelledMessage;

// UI tree structure from Android accessibility service

export interface UiNode {
  id: string;
  text: string | null;
  contentDescription: string | null;
  className: string;
  clickable: boolean;
  enabled: boolean;
  bounds: {
    left: number;
    top: number;
    right: number;
    bottom: number;
  };
}

export interface UiTree {
  screen: {
    packageName: string;
    timestamp: number;
  };
  nodes: UiNode[];
}

// Triage result from LLM

export interface TriageResult {
  needsScreenshot: boolean;
  reason: string;
  requestSpeech: string;
}

// Visual analysis result from LLM

export interface AnalysisResult {
  answer: string;
  highlights: Highlight[];
}
